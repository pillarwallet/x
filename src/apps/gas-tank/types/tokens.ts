export type SelectedToken = {
  name: string;
  symbol: string;
  usdValue: string;
  logo: string;
  dailyPriceChange: number;
  chainId: number;
  decimals: number;
  address: string;
};

export type PayingToken = {
  name: string;
  symbol: string;
  logo: string;
  actualBal: string;
  totalUsd: number;
  totalRaw: string;
  chainId: number;
  address: string;
};

export enum SearchType {
  MyHoldings = 'My Holdings',
}

export enum SortType {
  Up = 'Up',
  Down = 'Down',
}
