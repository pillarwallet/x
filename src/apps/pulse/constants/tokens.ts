export const STABLE_CURRENCIES = [
  { chainId: 10, address: '0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85' }, // USDC on Optimism
  { chainId: 137, address: '0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359' }, // USDC on Polygon
  { chainId: 8453, address: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913' }, // USDC on Base
  { chainId: 42161, address: '0xaf88d065e77c8cC2239327C5EDb3A432268e5831' }, // USDC on Arbitrum
  { chainId: 56, address: '0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d' }, // USDC on BNB Smart Chain
  { chainId: 100, address: '0x2a22f9c3b484c3629090FeED35F17Ff8F88f76F0' }, // USDC on Gnosis
];
