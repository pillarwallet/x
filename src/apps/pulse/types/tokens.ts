type ParsedToken = {
  name: string;
  symbol: string;
  markrtCap: number;
  liquidity: number;
  volume: number;
  logo: string;
  price: number;
  priceChange: number;
  blockchain: string;
}

type SelectedToken = {
  name: string;
  symbol: string;
  usdValue: string;
  logo: string;
  dailyPriceChange: number;
  chainId: number;
  decimals: number;
  address: string;
}