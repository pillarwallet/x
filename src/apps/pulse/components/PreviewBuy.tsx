import { ExpressIntentResponse } from "@etherspot/intent-sdk/dist/cjs/sdk/types/user-intent-types";
import Esc from "./Esc";
import PayingToken  from "./PayingToken";
import Refresh from "./Refresh";
import useIntentSdk from "../hooks/useIntentSdk";
import { useState } from "react";
import { TailSpin } from "react-loader-spinner";
import IntentTracker from "./IntentTracker";
import { getLogoForChainId } from "../../../utils/blockchain";

interface Props {
  closePreview: () => void,
  buyToken?: SelectedToken | null,
  payingTokens: PayingToken[],
  expressIntentResponse: ExpressIntentResponse | null
}

export default function PreviewBuy(props: Props) {
  const totalPay = props.payingTokens.reduce((acc, curr) => acc + curr.totalUsd, 0).toFixed(2);

  const [isLoading, setIsLoading] = useState(false);
  const [showTracker, setShowTracker] = useState(false);

  const detailsEntry = (lhs: string, rhs: string, moreInfo = true, tokenName = '') => {
    return (
      <div className="flex justify-between" style={{margin: 10}}>
        <div className="flex" style={{color: "grey", fontSize: 13}}>
          <div>{lhs}</div>
          {
            moreInfo &&
            <div style={{marginTop: 4, marginLeft: 4}}>
              <svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="12" height="12" rx="6" fill="white" fill-opacity="0.1"/>
                <g opacity="0.5">
                  <path d="M8.08335 4.75002C8.08335 4.19749 7.86386 3.66758 7.47316 3.27688C7.08246 2.88618 6.55255 2.66669 6.00002 2.66669C5.44749 2.66669 4.91758 2.88618 4.52688 3.27688C4.13618 3.66758 3.91669 4.19749 3.91669 4.75002C3.91669 4.86053 3.96059 4.96651 4.03873 5.04465C4.11687 5.12279 4.22285 5.16669 4.33335 5.16669C4.44386 5.16669 4.54984 5.12279 4.62798 5.04465C4.70612 4.96651 4.75002 4.86053 4.75002 4.75002C4.75002 4.50279 4.82333 4.26112 4.96068 4.05556C5.09804 3.85 5.29326 3.68978 5.52167 3.59517C5.75007 3.50056 6.00141 3.47581 6.24388 3.52404C6.48636 3.57227 6.70909 3.69132 6.8839 3.86614C7.05872 4.04095 7.17777 4.26368 7.226 4.50616C7.27423 4.74863 7.24948 4.99997 7.15487 5.22837C7.06026 5.45678 6.90004 5.65201 6.69448 5.78936C6.48892 5.92671 6.24725 6.00002 6.00002 6.00002C5.88951 6.00002 5.78353 6.04392 5.70539 6.12206C5.62725 6.2002 5.58335 6.30618 5.58335 6.41669V7.25002C5.58335 7.36053 5.62725 7.46651 5.70539 7.54465C5.78353 7.62279 5.88951 7.66669 6.00002 7.66669C6.11053 7.66669 6.21651 7.62279 6.29465 7.54465C6.37279 7.46651 6.41669 7.36053 6.41669 7.25002V6.79169C6.88726 6.69563 7.31019 6.43994 7.6139 6.0679C7.91761 5.69585 8.08345 5.23029 8.08335 4.75002Z" fill="white"/>
                  <path d="M6.00016 9.33327C6.23028 9.33327 6.41683 9.14672 6.41683 8.91661C6.41683 8.68649 6.23028 8.49994 6.00016 8.49994C5.77004 8.49994 5.5835 8.68649 5.5835 8.91661C5.5835 9.14672 5.77004 9.33327 6.00016 9.33327Z" fill="white"/>
                </g>
              </svg>
            </div>
          }
        </div>
        <div>
          <div className="flex" style={{fontSize: 13}}>
            <div>{rhs}</div>
            {
              tokenName &&
              <div style={{color: "grey"}}>{tokenName}</div>
            }
          </div>
        </div>
      </div>
    )
  }

  const {intentSdk} = useIntentSdk();

  const shortlistBid = async () => {
    setIsLoading(true);
    try {
      const res = await intentSdk?.shortlistBid(
        props.expressIntentResponse?.intentHash!, props.expressIntentResponse?.bids[0].bidHash!
      );
      setShowTracker(true);
    } catch (error) {
      console.log("shortlisting bid failed:: ", error);
    }
    setIsLoading(false);
  }

  return (
    <div
      className="flex flex-col"
      style={{
        border: '2px solid #1E1D24',
        width: 446,
        backgroundColor: "#121116",
        borderRadius: 10,
      }}
    >
      {
        showTracker ?
        <>
        <IntentTracker
          closePreview={props.closePreview}
          bidHash={props.expressIntentResponse?.bids[0].bidHash!}
          token={props.buyToken!}
          isBuy={true}
          amount={props.buyToken?.usdValue ? (Number(totalPay)/Number(props.buyToken.usdValue)).toFixed(4) : ""}
        />
        </> :
        <>
        <div className="flex justify-between" style={{margin: 10}}>
          <div style={{fontSize: 20}}>Preview</div>
          <div className="flex">
            <Refresh />
            <div style={{marginLeft: 10}}>
              <Esc closePreview={props.closePreview} />
            </div>
          </div>
        </div>

        <div className="flex justify-between" style={{margin: 10, marginBottom: 0, fontSize: 13, color: "grey"}}>
          <div>You’re paying</div>
          <div className="flex">Total: ${totalPay}</div>
        </div>

        <div className="rounded-[10px]" style={{width: 422, minHeight: 50, background: "black", margin: 10}}>
          {props.payingTokens.map((item) => <PayingToken payingToken={item}/>)}
        </div>

        <div className="flex justify-between" style={{margin: 10, marginBottom: 0, fontSize: 13, color: "grey"}}>
          <div>You’re buying</div>
        </div>

        <div>
          <div
            className="flex justify-between"
            style={{width: 422, height: 50, background: "black", borderRadius: 10, margin: 10}}
          >
            <div className="flex items-center">
              <div style={{position: "relative", display: "inline-block" }}>
                <img
                  src={props.buyToken?.logo}
                  alt="Main"
                  style={{
                    width: 32,
                    height: 32,
                    borderRadius: 50,
                    marginLeft: 5,
                    marginRight: 5
                  }}
                />
                <img
                  src={getLogoForChainId(props.buyToken?.chainId!)}
                  style={{position: "absolute",
                    bottom: "-1px",
                    right: "2px",
                    width: 12,
                    height: 12,
                    borderRadius: "50%",
                  }}
                />
              </div>
              <div style={{marginLeft: 5}}>
                <div className="flex">
                  <div style={{fontSize: 13}}>{props.buyToken?.name}</div>
                  <div style={{color: "grey", marginLeft: 5, fontSize: 13}}>{props.buyToken?.symbol}</div>
                </div>
                <div className="flex">
                  <div style={{fontSize: 13, color: "grey"}}>{props.buyToken?.address?.slice(0,6)}...{props.buyToken?.address.slice(-4)}</div>
                  <button style={{marginLeft: 5}} onClick={() => {navigator.clipboard.writeText(props.buyToken?.address ?? '')}}>
                    <svg width="10" height="12" viewBox="0 0 10 12" fill="none" xmlns="http://www.w3.org/2000/svg">
                      <g opacity="0.5">
                        <path opacity="0.5" d="M4 1.50004C4 0.671611 4.67157 3.76701e-05 5.5 3.76701e-05H8.5C9.32843 3.76701e-05 10 0.671611 10 1.50004V6.50004C10 7.32846 9.32843 8.00004 8.5 8.00004H7V5C7 3.89543 6.10457 3 5 3H4L4 1.50004Z" fill="white"/>
                        <path d="M0 5.5C0 4.67157 0.671573 4 1.5 4H4.5C5.32843 4 6 4.67157 6 5.5V10.5C6 11.3284 5.32843 12 4.5 12H1.5C0.671573 12 0 11.3284 0 10.5V5.5Z" fill="white"/>
                      </g>
                    </svg>
                  </button>
                </div>
              </div>
            </div>
            <div className="flex flex-col justify-center" style={{marginRight: 10}}>
              <div className="flex" style={{fontSize: 13, textAlign: "right"}}>{props.buyToken?.usdValue ? (Number(totalPay)/Number(props.buyToken.usdValue)).toFixed(4) : ""}</div>
              <div className="flex justify-end" style={{fontSize: 12, color: "grey", textAlign: "right"}}>${totalPay}</div>
            </div>
          </div>
        </div>

        <div className="flex justify-between" style={{margin: 10, marginBottom: 0, fontSize: 13, color: "grey"}}>
          <div>Details</div>
        </div>

        <div>
          <div style={{width: 422, height: 137, background: "black", borderRadius: 10, margin: 10}}>
            {
              detailsEntry("Rate", `1 USD ≈ ${props.buyToken?.usdValue ? Number(1/Number(props.buyToken.usdValue)).toFixed(3) : 1.000}`, false, props.buyToken?.symbol ?? '')
            }
            {
              detailsEntry("Minimum Receive", totalPay)
            }
            {
              detailsEntry("Price Impact", "0.00%")
            }
            {
              detailsEntry("Max Spillage", "0.0%")
            }
            {
              detailsEntry("Gas Fee", "≈ $0.00")
            }
          </div>
        </div>
        
        <div
          className="flex"
          style={{margin: 10, width: 422, height: 50, borderRadius: 10, backgroundColor: "black"}}
        >
          <button
            className="flex-1 items-center justify-center"
            style={{
              borderRadius: 10,
              backgroundColor: isLoading ? "#121116" : "#8A77FF",
              margin: 2
            }}
            onClick={shortlistBid}
            disabled={isLoading}
          >
            {
              isLoading  ?
              <div className="flex items-center justify-center">
                Waiting for signature...
              </div> :
              <>Confirm</>
            }
          </button>
        </div>
        </>
      }
      
    </div>
  )
}