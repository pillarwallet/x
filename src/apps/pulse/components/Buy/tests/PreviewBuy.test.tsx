/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { ExpressIntentResponse } from '@etherspot/intent-sdk/dist/cjs/sdk/types/user-intent-types';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// hooks
import useIntentSdk from '../../../hooks/useIntentSdk';

// types
import { PayingToken, SelectedToken } from '../../../types/tokens';

// components
import PreviewBuy from '../PreviewBuy';

// Mock dependencies
vi.mock('../../../hooks/useIntentSdk', () => ({
  default: vi.fn(),
}));

vi.mock('../../../../utils/blockchain', () => ({
  getLogoForChainId: vi.fn(() => '/src/assets/images/logo-ethereum.png'),
}));

const mockToken: SelectedToken = {
  name: 'Test Token',
  symbol: 'TEST',
  logo: 'test-logo.png',
  usdValue: '100.00',
  dailyPriceChange: 0.05,
  chainId: 1,
  decimals: 18,
  address: '0x1234567890123456789012345678901234567890',
};

const mockPayingToken: PayingToken = {
  name: 'USD Coin',
  symbol: 'USDC',
  logo: 'usdc-logo.png',
  actualBal: '100.00',
  totalUsd: 100.0,
  totalRaw: '100000000',
  chainId: 1,
  address: '0xA0b86a33E6441b8C4C8C0C4C8C0C4C8C0C4C8C0C4',
};

const mockExpressIntentResponse: ExpressIntentResponse = {
  intentHash: '0xIntentHash123456789',
  bids: [{ bidHash: '0xBidHash123456789' }],
} as any;

const mockProps = {
  closePreview: vi.fn(),
  buyToken: mockToken,
  payingTokens: [mockPayingToken],
  expressIntentResponse: mockExpressIntentResponse,
  setExpressIntentResponse: vi.fn(),
  usdAmount: '100.00',
  dispensableAssets: [],
};

const defaultMocks = () => {
  (useIntentSdk as any).mockReturnValue({
    intentSdk: {
      shortlistBid: vi.fn().mockResolvedValue(undefined),
    },
  });
};

describe('<PreviewBuy />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    defaultMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer.create(<PreviewBuy {...mockProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('renders preview interface', () => {
    it('displays main elements correctly', () => {
      render(<PreviewBuy {...mockProps} />);

      expect(screen.getByText('Confirm Transaction')).toBeInTheDocument();
      expect(screen.getAllByText('USD Coin')).toHaveLength(1);
      expect(screen.getAllByText('TEST')).toHaveLength(2);
      // Check that the component renders the main interface elements
      expect(screen.getByText('Total: $100.00')).toBeInTheDocument();
      expect(screen.getByText('Details')).toBeInTheDocument();
      expect(screen.getByText('Confirm')).toBeInTheDocument();
    });

    it('displays transaction details', () => {
      render(<PreviewBuy {...mockProps} />);

      expect(screen.getByText('Rate')).toBeInTheDocument();
      expect(screen.getByText('Minimum receive')).toBeInTheDocument();
      expect(screen.getByText('Price impact')).toBeInTheDocument();
      expect(screen.getByText('Max slippage')).toBeInTheDocument();
      expect(screen.getByText('Gas fee')).toBeInTheDocument();
    });

    it('shows token address with copy functionality', () => {
      render(<PreviewBuy {...mockProps} />);

      expect(screen.getByText('0x1234...7890')).toBeInTheDocument();
      expect(screen.getAllByAltText('copy-address-icon')).toHaveLength(2);
    });

    it('calculates token amount correctly', () => {
      render(<PreviewBuy {...mockProps} />);

      // 100 USD / 100 USD per token = 1 token
      expect(screen.getByText('1.000000')).toBeInTheDocument();
    });
  });

  describe('handles user interactions', () => {
    it('executes copy functionality', async () => {
      render(<PreviewBuy {...mockProps} />);

      // Check that the copy buttons exist
      const copyButtons = screen.getAllByAltText('copy-address-icon');
      expect(copyButtons).toHaveLength(2);

      // Click the first copy button - this should not throw an error
      expect(() => fireEvent.click(copyButtons[0])).not.toThrow();
    });

    it('executes shortlist bid transaction', async () => {
      const mockShortlistBid = vi.fn().mockResolvedValue(undefined);
      (useIntentSdk as any).mockReturnValue({
        intentSdk: {
          shortlistBid: mockShortlistBid,
        },
      });

      render(<PreviewBuy {...mockProps} />);

      const confirmButton = screen.getByText('Confirm');
      fireEvent.click(confirmButton);

      await waitFor(() => {
        expect(
          screen.getByText(
            'Please open your wallet and confirm the transaction.'
          )
        ).toBeInTheDocument();
      });

      expect(mockShortlistBid).toHaveBeenCalledWith(
        mockExpressIntentResponse.intentHash,
        mockExpressIntentResponse.bids[0].bidHash
      );
    });

    it('handles close functionality', () => {
      render(<PreviewBuy {...mockProps} />);

      const closeButton = screen.getByLabelText('Close');
      fireEvent.click(closeButton);

      expect(mockProps.closePreview).toHaveBeenCalled();
    });

    it('shows intent tracker after successful shortlist', async () => {
      render(<PreviewBuy {...mockProps} />);

      const confirmButton = screen.getByText('Confirm');
      fireEvent.click(confirmButton);

      await waitFor(() => {
        expect(
          screen.getByText(
            'Please open your wallet and confirm the transaction.'
          )
        ).toBeInTheDocument();
      });

      // After shortlist completes, should show tracker
      await waitFor(() => {
        expect(screen.getByText('Resource Lock Creation')).toBeInTheDocument();
      });
    });
  });

  describe('handles error states', () => {
    it('handles shortlist bid failure', async () => {
      (useIntentSdk as any).mockReturnValue({
        intentSdk: {
          shortlistBid: vi
            .fn()
            .mockRejectedValue(new Error('Shortlist failed')),
        },
      });

      render(<PreviewBuy {...mockProps} />);

      const confirmButton = screen.getByText('Confirm');
      fireEvent.click(confirmButton);

      await waitFor(() => {
        expect(
          screen.getByText(
            'Please open your wallet and confirm the transaction.'
          )
        ).toBeInTheDocument();
      });

      // Should not show tracker on error
      expect(
        screen.queryByText(
          'Please open your wallet and confirm the transaction.'
        )
      ).not.toBeInTheDocument();
    });

    it('handles missing buy token', () => {
      render(<PreviewBuy {...mockProps} buyToken={null} />);

      expect(
        screen.getByText(
          'No buy offer was found. Please check the token and the input amount and try again.'
        )
      ).toBeInTheDocument();
      expect(screen.queryByText('Test Token')).not.toBeInTheDocument();
    });

    it('handles missing express intent response', () => {
      render(<PreviewBuy {...mockProps} expressIntentResponse={null} />);

      expect(
        screen.getByText(
          'No buy offer was found. Please check the token and the input amount and try again.'
        )
      ).toBeInTheDocument();
      expect(screen.getByText('Close')).toBeInTheDocument();
    });
  });

  describe('handles edge cases', () => {
    it('handles token without logo', () => {
      const tokenWithoutLogo = { ...mockToken, logo: '' };
      render(<PreviewBuy {...mockProps} buyToken={tokenWithoutLogo} />);

      expect(screen.getByTestId('random-avatar')).toBeInTheDocument();
    });

    it('handles multiple paying tokens', () => {
      const multiplePayingTokens = [
        { ...mockPayingToken, totalUsd: 50.0 },
        { ...mockPayingToken, name: 'Dai', symbol: 'DAI', totalUsd: 50.0 },
      ];

      render(<PreviewBuy {...mockProps} payingTokens={multiplePayingTokens} />);

      expect(screen.getByText('Total: $100.00')).toBeInTheDocument();
    });

    it('handles zero total payment', () => {
      const zeroPayingTokens = [{ ...mockPayingToken, totalUsd: 0 }];

      render(<PreviewBuy {...mockProps} payingTokens={zeroPayingTokens} />);

      expect(screen.getByText('Total: $0.00')).toBeInTheDocument();
    });

    it('handles token with zero USD value', () => {
      const tokenWithZeroValue = { ...mockToken, usdValue: '0' };

      render(<PreviewBuy {...mockProps} buyToken={tokenWithZeroValue} />);

      // Just check that the component renders without crashing
      expect(screen.getByText('Confirm Transaction')).toBeInTheDocument();
    });

    it('handles missing token address', () => {
      const tokenWithoutAddress = { ...mockToken, address: '' };

      render(<PreviewBuy {...mockProps} buyToken={tokenWithoutAddress} />);

      expect(screen.getByText('Address not available')).toBeInTheDocument();
    });
  });

  describe('displays correct calculations', () => {
    it('calculates rate correctly', () => {
      render(<PreviewBuy {...mockProps} />);

      // 1 USD ≈ 0.01 TEST (1 / 100)
      expect(screen.getByText('1 USD ≈ 0.010')).toBeInTheDocument();
    });

    it('displays minimum receive as total payment', () => {
      render(<PreviewBuy {...mockProps} />);

      expect(screen.getAllByText('$100.00')).toHaveLength(2);
    });

    it('shows correct token amount calculation', () => {
      const customPayingTokens = [{ ...mockPayingToken, totalUsd: 250.0 }];
      const customToken = { ...mockToken, usdValue: '50.00' };

      render(
        <PreviewBuy
          {...mockProps}
          buyToken={customToken}
          payingTokens={customPayingTokens}
        />
      );

      // 250 USD / 50 USD per token = 5 tokens
      expect(screen.getByText('5.000000')).toBeInTheDocument();
    });
  });
});
