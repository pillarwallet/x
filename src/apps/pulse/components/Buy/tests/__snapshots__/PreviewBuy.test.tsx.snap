// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PreviewBuy /> > renders correctly and matches snapshot 1`] = `
<div
  className="flex flex-col w-full max-w-[446px] bg-[#1E1D24] border border-white/5 rounded-[10px] p-6"
>
  <div
    className="flex justify-between mb-6"
  >
    <div
      className="text-xl font-normal"
    >
      Confirm Transaction
    </div>
    <div
      className="flex"
    >
      <div
        className="bg-[#121116] rounded-[10px] w-10 h-10 p-[2px_2px_4px_2px]"
      >
        <button
          aria-label="Refresh"
          className="flex items-center justify-center w-[36px] h-[34px] bg-[#1E1D24] rounded-[8px] cursor-pointer"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <img
            alt="refresh-icon"
            src="data:image/svg+xml,%3csvg%20width='20'%20height='18'%20viewBox='0%200%2020%2018'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cg%20opacity='0.5'%3e%3cpath%20d='M1.93781%209.03579C1.94148%207.2726%202.52824%205.55746%203.6071%204.1564C4.68595%202.75533%206.19659%201.74663%207.90475%201.28674C9.6129%200.826854%2011.4231%200.941474%2013.0546%201.61283C14.6861%202.28419%2016.0477%203.47475%2016.9282%204.99989M16.9282%204.99989L13.7631%204.51784M16.9282%204.99989L16.7272%201.65181'%20stroke='white'%20stroke-width='2'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3cpath%20opacity='0.5'%20d='M18.0619%208.96372C18.0583%2010.7269%2017.4715%2012.442%2016.3927%2013.8431C15.3138%2015.2442%2013.8032%2016.2529%2012.095%2016.7128C10.3869%2017.1727%208.57666%2017.058%206.94518%2016.3867C5.31369%2015.7153%203.9521%2014.5248%203.07156%2012.9996M3.07156%2012.9996L6.23662%2013.4817M3.07156%2012.9996L3.27252%2016.3477'%20stroke='white'%20stroke-width='2'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3c/g%3e%3c/svg%3e"
          />
        </button>
      </div>
      <div
        className="bg-[#121116] rounded-[10px] w-10 h-10 p-[2px_2px_4px_2px] ml-[10px]"
      >
        <button
          aria-label="Close"
          className="flex items-center justify-center w-[36px] h-[34px] bg-[#1E1D24] rounded-[8px]"
          data-testid="esc-button"
          onClick={[MockFunction spy]}
          type="button"
        >
          <p
            className="text-white/50 font-medium text-[13px] leading-[100%] tracking-[-0.02em] text-center"
          >
            ESC
          </p>
        </button>
      </div>
    </div>
  </div>
  <div
    className="flex justify-between mb-3 text-[13px] font-normal text-white/50"
  >
    <div>
      You're paying
    </div>
    <div
      className="flex"
    >
      Total: $
      100.00
    </div>
  </div>
  <div
    className="flex flex-col w-full border border-[#25232D] rounded-[10px] p-3 mb-6"
  >
    <div
      className="flex justify-between"
    >
      <div
        className="flex items-center"
      >
        <div
          className="relative inline-block mr-2"
        >
          <img
            alt="Main"
            className="w-8 h-8 rounded-full"
            src="usdc-logo.png"
          />
          <img
            alt="Chain logo"
            className="absolute -bottom-px right-0.5 w-3 h-3 rounded-full border border-[#1E1D24]"
            src="/src/assets/images/logo-ethereum.png"
          />
        </div>
        <div>
          <div
            className="flex items-center text-[13px] font-normal text-white"
          >
            <span>
              USD Coin
            </span>
            <span
              className="ml-1 text-white/50"
            >
              USDC
            </span>
          </div>
          <div
            className="flex items-center text-[13px] font-normal text-white/50"
          >
            <span>
              0xA0b8...C0C4
            </span>
            <div
              className="flex items-center ml-1 cursor-pointer"
              onClick={[Function]}
            >
              <img
                alt="copy-address-icon"
                className="w-[10px] h-3"
                src="data:image/svg+xml,%3csvg%20width='10'%20height='12'%20viewBox='0%200%2010%2012'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cg%20opacity='0.5'%3e%3cpath%20opacity='0.5'%20d='M4%201.50004C4%200.671611%204.67157%203.76701e-05%205.5%203.76701e-05H8.5C9.32843%203.76701e-05%2010%200.671611%2010%201.50004V6.50004C10%207.32846%209.32843%208.00004%208.5%208.00004H7V5C7%203.89543%206.10457%203%205%203H4L4%201.50004Z'%20fill='white'/%3e%3cpath%20d='M0%205.5C0%204.67157%200.671573%204%201.5%204H4.5C5.32843%204%206%204.67157%206%205.5V10.5C6%2011.3284%205.32843%2012%204.5%2012H1.5C0.671573%2012%200%2011.3284%200%2010.5V5.5Z'%20fill='white'/%3e%3c/g%3e%3c/svg%3e"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="flex flex-col justify-center text-right"
      >
        <div
          className="text-[13px] font-normal text-white"
        >
          100000000
        </div>
        <div
          className="text-xs font-normal text-white/50"
        >
          $
          100.00
        </div>
      </div>
    </div>
  </div>
  <div
    className="flex justify-between mb-3 text-[13px] font-normal text-white/50"
  >
    <div>
      You're buying
    </div>
  </div>
  <div
    className="flex justify-between w-full border border-[#25232D] rounded-[10px] p-3 mb-6"
  >
    <div
      className="flex items-center"
    >
      <div
        className="relative inline-block mr-2"
      >
        <img
          alt="Main"
          className="w-8 h-8 rounded-full"
          src="test-logo.png"
        />
        <img
          alt="Chain logo"
          className="absolute -bottom-px right-0.5 w-3 h-3 rounded-full border border-[#1E1D24]"
          src="/src/assets/images/logo-ethereum.png"
        />
      </div>
      <div>
        <div
          className="flex items-center text-[13px] font-normal text-white"
        >
          <span>
            USD Coin
          </span>
          <span
            className="ml-1 text-white/50"
          >
            TEST
          </span>
        </div>
        <div
          className="flex items-center text-[13px] font-normal text-white/50"
        >
          <span>
            0x1234...7890
          </span>
          <div
            className="flex items-center ml-1 cursor-pointer"
            onClick={[Function]}
          >
            <img
              alt="copy-address-icon"
              className="w-[10px] h-3"
              src="data:image/svg+xml,%3csvg%20width='10'%20height='12'%20viewBox='0%200%2010%2012'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cg%20opacity='0.5'%3e%3cpath%20opacity='0.5'%20d='M4%201.50004C4%200.671611%204.67157%203.76701e-05%205.5%203.76701e-05H8.5C9.32843%203.76701e-05%2010%200.671611%2010%201.50004V6.50004C10%207.32846%209.32843%208.00004%208.5%208.00004H7V5C7%203.89543%206.10457%203%205%203H4L4%201.50004Z'%20fill='white'/%3e%3cpath%20d='M0%205.5C0%204.67157%200.671573%204%201.5%204H4.5C5.32843%204%206%204.67157%206%205.5V10.5C6%2011.3284%205.32843%2012%204.5%2012H1.5C0.671573%2012%200%2011.3284%200%2010.5V5.5Z'%20fill='white'/%3e%3c/g%3e%3c/svg%3e"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex flex-col justify-center text-right"
    >
      <div
        className="text-[13px] font-normal text-white"
      >
        1.000000
      </div>
      <div
        className="text-xs font-normal text-white/50"
      >
        $
        100.00
      </div>
    </div>
  </div>
  <div
    className="flex justify-between mb-3 text-[13px] font-normal text-white/50"
  >
    <div>
      Details
    </div>
  </div>
  <div
    className="flex flex-col w-full border border-[#25232D] rounded-[10px] p-3 mb-3"
  >
    <div
      className="flex justify-between mb-3"
    >
      <div
        className="flex items-center text-white/50 text-[13px] font-normal"
      >
        <div>
          Rate
        </div>
        <div
          className="ml-1.5"
        >
          <div
            className="relative inline-block"
          >
            <div
              className="cursor-help"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="w-3 h-3 rounded-full bg-white/10 flex items-center justify-center"
              >
                <span
                  className="text-white/50 text-[10px] font-normal"
                >
                  ?
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="flex items-center text-[13px] font-normal text-white"
        >
          <div>
            1 USD â‰ˆ 0.010
          </div>
          <div
            className="ml-1 text-white/50"
          >
            TEST
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex justify-between mb-3"
    >
      <div
        className="flex items-center text-white/50 text-[13px] font-normal"
      >
        <div>
          Minimum receive
        </div>
        <div
          className="ml-1.5"
        >
          <div
            className="relative inline-block"
          >
            <div
              className="cursor-help"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="w-3 h-3 rounded-full bg-white/10 flex items-center justify-center"
              >
                <span
                  className="text-white/50 text-[10px] font-normal"
                >
                  ?
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="flex items-center text-[13px] font-normal text-white"
        >
          <div>
            0.970000 TEST
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex justify-between mb-3"
    >
      <div
        className="flex items-center text-white/50 text-[13px] font-normal"
      >
        <div>
          Price impact
        </div>
        <div
          className="ml-1.5"
        >
          <div
            className="relative inline-block"
          >
            <div
              className="cursor-help"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="w-3 h-3 rounded-full bg-white/10 flex items-center justify-center"
              >
                <span
                  className="text-white/50 text-[10px] font-normal"
                >
                  ?
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="flex items-center text-[13px] font-normal text-white"
        >
          <div>
            0.00%
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex justify-between mb-3"
    >
      <div
        className="flex items-center text-white/50 text-[13px] font-normal"
      >
        <div>
          Max slippage
        </div>
        <div
          className="ml-1.5"
        >
          <div
            className="relative inline-block"
          >
            <div
              className="cursor-help"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="w-3 h-3 rounded-full bg-white/10 flex items-center justify-center"
              >
                <span
                  className="text-white/50 text-[10px] font-normal"
                >
                  ?
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="flex items-center text-[13px] font-normal text-white"
        >
          <div>
            3.0%
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex justify-between mb-3"
    >
      <div
        className="flex items-center text-white/50 text-[13px] font-normal"
      >
        <div>
          Gas fee
        </div>
        <div
          className="ml-1.5"
        >
          <div
            className="relative inline-block"
          >
            <div
              className="cursor-help"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="w-3 h-3 rounded-full bg-white/10 flex items-center justify-center"
              >
                <span
                  className="text-white/50 text-[10px] font-normal"
                >
                  ?
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="flex items-center text-[13px] font-normal text-white"
        >
          <div>
            â‰ˆ $0.00
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="w-full rounded-[10px] bg-[#121116] p-[2px_2px_6px_2px]"
  >
    <button
      className="flex items-center justify-center w-full rounded-[8px] h-[42px] p-[1px_6px_1px_6px] bg-[#8A77FF]"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Confirm
    </button>
  </div>
</div>
`;
